File: demo/src/main/java/com/example/ConsumidorInternacional.java
package com.example;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.DeliverCallback;

/**
 * Consumidor de Pedidos INTERNACIONALES
 *
 * Se suscribe a mensajes con routing key "pedido.internacional".
 * Simula el flujo aduanero y logistico de envios al exterior (15-30 dias).
 */
public class ConsumidorInternacional {

    private static final String EXCHANGE_NAME = "pedidos_exchange";
    private static final String QUEUE_NAME = "cola_internacional";
    private static final String ROUTING_KEY = "pedido.internacional";

    public static void main(String[] args) throws Exception {

        // 1. Conexion a RabbitMQ
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("localhost");
        factory.setPort(5672);

        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();

        // 2. Declarar exchange topic
        channel.exchangeDeclare(EXCHANGE_NAME, "topic");

        // 3. Declarar cola y binding con routing key internacional
        channel.queueDeclare(QUEUE_NAME, true, false, false, null);
        channel.queueBind(QUEUE_NAME, EXCHANGE_NAME, ROUTING_KEY);

        System.out.println("[ConsumidorInternacional] Esperando pedidos internacionales. CTRL+C para salir.");

        // 4. Callback de procesamiento
        DeliverCallback deliverCallback = (consumerTag, delivery) -> {
            String mensaje = new String(delivery.getBody(), "UTF-8");
            System.out.println("\n[ConsumidorInternacional] Pedido recibido: " + mensaje);
            procesarPedidoInternacional(mensaje);
        };

        // 5. Iniciar consumo
        channel.basicConsume(QUEUE_NAME, true, deliverCallback, consumerTag -> {
        });
    }

    /**
     * Simula el flujo internacional: revision de restricciones de exportacion,
     * declaracion aduanal y contratacion de operador logistico internacional.
     */
    private static void procesarPedidoInternacional(String mensaje) {
        System.out.println("  -> [INTERNACIONAL] Revisando restricciones de exportacion...");
        System.out.println("  -> [INTERNACIONAL] Generando declaracion aduanal...");
        System.out.println("  -> [INTERNACIONAL] Calculando aranceles e impuestos...");
        System.out.println("  -> [INTERNACIONAL] Contratando operador logistico internacional...");
        System.out.println("  -> [INTERNACIONAL] Tiempo estimado de entrega: 15-30 dias. HECHO.");
    }
}

File: demoPy/consumidor_internacional.py
import pika
import json

"""
Consumidor de Pedidos INTERNACIONALES - Python
================================================
Se suscribe al exchange 'pedidos_exchange' con la routing key 'pedido.internacional'.
Gestiona el flujo aduanero y logistico de envios al exterior (15-30 dias).
"""

EXCHANGE_NAME = "pedidos_exchange"
QUEUE_NAME    = "cola_internacional"
ROUTING_KEY   = "pedido.internacional"


def procesar_pedido_internacional(pedido: dict):
    """
    Simula el flujo internacional: revision aduanal,
    calculo de aranceles y asignacion de operador logistico.
    """
    print(f"  -> [INTERNACIONAL] Revisando restricciones de exportacion para pedido #{pedido['id']}...")
    print(f"  -> [INTERNACIONAL] Generando declaracion aduanal para '{pedido['producto']}'...")
    print(f"  -> [INTERNACIONAL] Calculando aranceles e impuestos...")
    print(f"  -> [INTERNACIONAL] Contratando operador logistico internacional...")
    print(f"  -> [INTERNACIONAL] Entrega estimada: 15-30 dias. HECHO.\n")


def callback(ch, method, properties, body):
    pedido = json.loads(body.decode("utf-8"))
    print(f"[ConsumidorInternacional] Pedido recibido: {json.dumps(pedido)}")
    procesar_pedido_internacional(pedido)


def main():
    connection = pika.BlockingConnection(pika.ConnectionParameters("localhost"))
    channel = connection.channel()

    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type="topic")

    channel.queue_declare(queue=QUEUE_NAME)
    channel.queue_bind(exchange=EXCHANGE_NAME, queue=QUEUE_NAME, routing_key=ROUTING_KEY)

    channel.basic_consume(queue=QUEUE_NAME, on_message_callback=callback, auto_ack=True)

    print("[ConsumidorInternacional] Esperando pedidos internacionales. CTRL+C para salir.")
    channel.start_consuming()


if __name__ == "__main__":
    main()
